sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala

```diff
@@ -0,0 +1,139 @@
+import org.apache.spark.sql.catalyst.expressions.aggregate.{Complete, Final, Partial}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.aggregate.HashAggregateExec
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * Replace [[HashAggregateExec]] with [[SortAggregateExec]] in the spark plan if:
+ *
+ * 1. The plan is a pair of partial and final [[HashAggregateExec]], and the child of partial
+ *    aggregate satisfies the sort order of corresponding [[SortAggregateExec]].
+ * or
+ * 2. The plan is a [[HashAggregateExec]], and the child satisfies the sort order of
+ *    corresponding [[SortAggregateExec]].
+ *
+ * Examples:
+ * 1. aggregate after join:
+ *
+ *  HashAggregate(t1.i, SUM, final)
+ *               |                         SortAggregate(t1.i, SUM, complete)
+ * HashAggregate(t1.i, SUM, partial)   =>                |
+ *               |                            SortMergeJoin(t1.i = t2.j)
+ *    SortMergeJoin(t1.i = t2.j)
+ *
+ * 2. aggregate after sort:
+ *
+ * HashAggregate(t1.i, SUM, partial)         SortAggregate(t1.i, SUM, partial)
+ *               |                     =>                  |
+ *           Sort(t1.i)                                Sort(t1.i)
+ *
+ * [[HashAggregateExec]] can be replaced when its child satisfies the sort order of
+ * corresponding [[SortAggregateExec]]. [[SortAggregateExec]] is faster in the sense that
+ * it does not have hashing overhead of [[HashAggregateExec]].
+ */
+object ReplaceHashWithSortAgg extends Rule[SparkPlan] {
+  def apply(plan: SparkPlan): SparkPlan = {
+    if (!conf.getConf(SQLConf.REPLACE_HASH_WITH_SORT_AGG_ENABLED)) {
+      plan
+    } else {
+      replaceHashAgg(plan)
+    }
+  }
+
+  /**
+   * Replace [[HashAggregateExec]] with [[SortAggregateExec]].
+   */
+  private def replaceHashAgg(plan: SparkPlan): SparkPlan = {
+    plan.transformDown {
+      case hashAgg: HashAggregateExec if hashAgg.groupingExpressions.nonEmpty =>
+        val sortAgg = hashAgg.toSortAggregate
+        hashAgg.child match {
+          case partialAgg: HashAggregateExec if isPartialAgg(partialAgg, hashAgg) =>
+            if (SortOrder.orderingSatisfies(
+                partialAgg.child.outputOrdering, sortAgg.requiredChildOrdering.head)) {
+              sortAgg.copy(
+                aggregateExpressions = sortAgg.aggregateExpressions.map(_.copy(mode = Complete)),
+                child = partialAgg.child)
+            } else {
+              hashAgg
+            }
+          case other =>
+            if (SortOrder.orderingSatisfies(
+                other.outputOrdering, sortAgg.requiredChildOrdering.head)) {
+              sortAgg
+            } else {
+              hashAgg
+            }
+        }
+      case other => other
+    }
+  }
+
+  /**
+   * Check if `partialAgg` to be partial aggregate of `finalAgg`.
+   */
+  private def isPartialAgg(partialAgg: HashAggregateExec, finalAgg: HashAggregateExec): Boolean = {
```

Comment:
This looks like reverse enginering the `AggUtils`. Could we just link the partial and final agg when they are constructed?

Reply:
@tanelk - yeah I agree this is mostly reverse engineering and we can do a better job here. I tried link partial and final agg in `AggUtils` and check linked physical plan to be same or not. This does not quite work due to we are doing top-down planning, and the linked partial agg not being same as planned partial agg (having `PlanLater` operator in linked partial agg).

I found a more elegant way to do it, by checking the linked logical plan of both aggs to be same. Updated.

Reply:
cc @cloud-fan for review, thanks.